core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/util/QueryEvaluationUtil.java

```diff
@@ -191,6 +191,31 @@ public static boolean compareLiterals(Literal leftLit, Literal rightLit, Compare
 				} else if (!strict && leftXsdDatatype.isDurationDatatype() && rightXsdDatatype.isDurationDatatype()) {
 					commonDatatype = XSD.DURATION;
 				}
+			} else if (commonDatatype == null && (leftXsdDatatype != null || rightXsdDatatype != null)) {
```

Comment:
I'm not following the logic here. The `else if` condition checks if `leftXsdDatatype` or `rightXsdDatatype` are not null, but then the body of this condition not do anything with the enums at all, and just uses `leftDatatype` and `rightDatatype` (the IRI values).

Reply:
(as an aside that compareLiterals method is a monster of nested conditionals - out of scope to refactor for a quick patch release of course but it could do with a bit of cleaning up)

Reply:
Might be a typo on my end. Wrote that code late last night. I'll fix it.

Reply:
Fixed